apps/examples/setjmp/setjmp_longjmp_test.c

```diff
@@ -0,0 +1,166 @@
+#include <tinyara/config.h>
+
+#include <assert.h>
+#include <setjmp.h>
+#include <stdio.h>
+#include <unistd.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define MAX_DEPTH 8
+int did_return[MAX_DEPTH];
+
+typedef enum {
+  DIVE_RETURN,
+  DIVE_LONGJMP,
+  DIVE_EXCEPTION
+} dive_mode_t;
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+// This function does a recursive dive MAX_DEPTH times, each time saving local
+// state on stack with setjmp call (ensure stck is big enough to hold them).
+//
+// Depending on the |mode| it either:
+//   DIVE_RETURN - just returns (dive all to the bottom, then return
+//                 all the way up, one frame at a time).
+//   DIVE_LONGJMP - calls down (until it reaches the bottom), then "returns"
+//                 via longjmp to its immediate caller, all the way up.
+//   DIVE_EXCEPTION - calls down (until it reaches the bottom), then returns
+//                 to the top, skipping intermediate stack frames.
+static void dive(int depth, dive_mode_t mode, jmp_buf upper) {
+  fprintf(stderr, "diving(%d, %c, %p)\n", depth, "RLE"[mode], (void *)upper);
+  jmp_buf env;
+
+  switch (mode) {
+  case DIVE_RETURN:
+    assert(setjmp(env) == 0);
+    if (depth < MAX_DEPTH) dive(depth + 1, mode, env);
+    break;
+  case DIVE_LONGJMP:
+    if ((setjmp(env) == 0) && (depth < MAX_DEPTH)) {
+      dive(depth + 1, mode, env);
+      assert(!"setjmp/longjmp failed: should not reach this point");
+    }
+    fprintf(stderr, "ascending(%d, %c, %p)\n", depth, "RLE"[mode], (void *)upper);
+    // 25 ms
+    usleep(25000);
+    did_return[depth - 1] = -1;
+    longjmp(upper, 1);
+    break;
+  case DIVE_EXCEPTION:
+    if ((setjmp(env) == 0) && (depth < MAX_DEPTH)) {
+      dive(depth + 1, mode, upper);
+    } else {
+      did_return[depth - 1] = -1;
+      fprintf(stderr, "ascending(%d, %c, %p)\n", depth, "RLE"[mode], (void *)upper);
+      longjmp(upper, 1);
+    }
+  }
+  did_return[depth - 1] = 1;
+  fprintf(stderr, "return(%d, %c, %p)\n", depth, "RLE"[mode], (void *)upper);
+  return;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int setjmp_test(int argc, char *argv[])
+#endif
+{
+  jmp_buf hook;
+  int i;
+
+  // Every depth of dive has returned.
+  for (i = 0; i < MAX_DEPTH; ++i) did_return[i] = 0;
+  assert(setjmp(hook) == 0);
```

Comment:
Do you want to use assert? instead of return ERROR?

Reply:
If setjmp/longjmp does not work, whole binary that makes calls to these functions is bad.
If you insist, I'll rework these to drop errors instead of crashing.

Should I?

Reply:
Then, you should change all of failure to assert in kernel side. To someone, it is not an important functionality and this is an one of examples. "return ERROR" is enough I think.